ma_gym/envs/traffic_junction/traffic_junction.py

```diff
@@ -125,25 +282,173 @@ def __create_grid(self):
 
         return _grid
 
+    def step(self, agents_action):
+        """
+        Performs an action in the environment and steps forward. At each step a new agent enters the road by
+        one of the 4 gates according to a probability "_arrive_prob". A "ncoll" reward is given to an agent if it
+        collides and all of them receive "-0.01*step_n" to avoid traffic jams.
+
+        :param agents_action: list of actions of all the agents to perform in the environment
+        :type agents_action: list
+
+        :return: agents observations, rewards, if agents are done and additional info
+        :rtype: tuple
+        """
+        assert len(agents_action) == self.n_agents, \
+            "Invalid action! It was expected to be list of {}" \
+            " dimension but was found to be of {}".format(self.n_agents, len(agents_action))
+
+        assert all([action_i in ACTION_MEANING.keys() for action_i in agents_action]), \
+            "Invalid action found in the list of sampled actions {}" \
+            ". Valid actions are {}".format(agents_action, ACTION_MEANING.keys())
+
+        self._step_count += 1
+        rewards = [0 for _ in range(self.n_agents)]  # initialize rewards array
+
+        # checks if there is a collision; this is done in the __update_agent_pos method
+        for agent_i, action in enumerate(agents_action):
+            if not (self._agent_dones[agent_i]) and self._on_the_road[agent_i]:
+                collision_flag = self.__update_agent_pos(agent_i, action)
+                if collision_flag:
+                    rewards[agent_i] += self._collision_reward
+
+            # checks if destination was reached
+            if self.__reached_dest(agent_i):
+                self._on_the_road[agent_i] = False
+                self.curr_cars_count -= 1
+
+            # Todo: Once a car reaches it's destination , it will never enter again in any of the tracks.
```

Comment:
Only new cars enter the grid and all those cars which have reached their destination wouldn't be considered for entering any track.

Also, if all cars have reached their destination, then the episode terminates.